src/tools/w3dviewer/w3dviewdoc.cpp

```diff
@@ -0,0 +1,957 @@
+/**
+ * @file
+ *
+ * @author Jonathan Wilson
+ *
+ * @brief W3D View document
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "w3dviewdoc.h"
+#include "assetmgr.h"
+#include "bmp2d.h"
+#include "camera.h"
+#include "datatreeview.h"
+#include "ffactory.h"
+#include "graphicview.h"
+#include "hanim.h"
+#include "light.h"
+#include "mainfrm.h"
+#include "mesh.h"
+#include "meshmdl.h"
+#include "part_emt.h"
+#include "resource.h"
+#include "texture.h"
+#include "utils.h"
+#include "viewerscene.h"
+#include "w3dview.h"
+
+IMPLEMENT_DYNCREATE(CW3DViewDoc, CDocument)
+
+// clang-format off
+BEGIN_MESSAGE_MAP(CW3DViewDoc, CDocument)
+END_MESSAGE_MAP()
+// clang-format on
+
+CW3DViewDoc::CW3DViewDoc() :
+    m_scene(nullptr),
+    m_textureScene(nullptr),
+    m_backgroundScene(nullptr),
+    m_model(nullptr),
+    m_backgroundObject(nullptr),
+    m_animation(nullptr),
+    m_light(nullptr),
+    m_backgroundBitmap(nullptr),
+    m_textureCamera(nullptr),
+    m_docCamera(nullptr),
+    m_fogColor(0.5f, 0.5f, 0.5f),
+    m_isQCompressed(false),
+    m_compressType(2),
+    m_frameCount(0.0f),
+    m_blendFrames(TRUE),
+    m_animateCamera(false),
+    m_resetCamera(true),
+    m_forceCameraReset(true),
+    m_useManualFov(false),
+    m_useManualClipPlanes(false),
+    m_initialized(false),
+    m_fogEnabled(false),
+    m_time(0.0f)
+{
+    m_animateCamera = theApp.GetProfileInt("Config", "AnimateCamera", 0) == 1;
+    m_resetCamera = theApp.GetProfileInt("Config", "ResetCamera", 1) == 1;
+}
+
+CW3DViewDoc::~CW3DViewDoc()
+{
+    Free();
+}
+
+BOOL CW3DViewDoc::OnNewDocument()
+{
+    BOOL result = CDocument::OnNewDocument();
+
+    if (result) {
+        m_loadedFiles.Delete_All();
+        m_forceCameraReset = true;
+
+        if (m_scene != nullptr && m_model != nullptr) {
+            RemoveRenderObject(m_model);
+        }
+
+        if (m_scene != nullptr) {
+            m_scene->Remove_All_LOD_Objects();
+            m_scene->Set_Fog_Color(m_fogColor);
+        }
+
+        if (m_model != nullptr) {

```

Comment:
With this test in place it looks like `m_animation` could leak. Maybe omit this test, because `Ref_Ptr_Release` already tests null?

Reply:
m_animation would only be set if m_model is also set.